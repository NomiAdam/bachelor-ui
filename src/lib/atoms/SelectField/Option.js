import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { resolveTheme, TYPES } from '../../utils/resolveTheme';

const StyledOption = styled.div`
    width: 100%;
    display: inline-block;
    border: none;
	  padding: 10px 20px;
	  text-align: center;
	  background-color: ${ ({ selected, ...props }) => (selected ? resolveTheme(TYPES.BACKGROUND_COLOR)(props) : 'initial') };
    &:hover {
        background-color: ${ resolveTheme(TYPES.BACKGROUND_COLOR_HOVER) };
        cursor: pointer;
    }
`;

const Option = ({
    value, onClick, label, selected, secondary,
}) => (
    <StyledOption secondary={ secondary } onMouseDown={ onClick(value) } selected={ selected }>
        {label}
    </StyledOption>
);

Option.propTypes = {
    /**
     * String value of Option
     */
    value: PropTypes.any,
    /**
     * onMouseDown event handler, (value) => (e) =>
     */
    onClick: PropTypes.func.isRequired,
    /**
     * String representation of displayed value
     */
    label: PropTypes.string,
    /**
     * Boolean representation of currently selected item
     */
    selected: PropTypes.bool,
};

export { StyledOption };
export default Option;
