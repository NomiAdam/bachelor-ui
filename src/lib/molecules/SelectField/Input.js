import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { lightTheme } from '../../constants/theme';
import { resolveTheme, TYPES } from '../../utils/resolveTheme';

const StyledInput = styled.input`
    border: none;
    border-bottom: 1px solid ${ resolveTheme(TYPES.COLOR) };
    box-sizing: border-box;
    width: 100%;
	  min-height: 30px;
	  padding: 10px 20px;
	  font-size: 16px;
	  display: inline-block;
	  text-align: center;
	  background-color: ${ ({ disabled }) => (disabled ? lightTheme.lightGrey : 'transparent') };
	  &:hover {
	    cursor: ${ ({ disabled }) => (disabled ? 'initial' : 'pointer') };
	  }
    &:focus {
        outline: none;
    }
`;

const Input = ({
    value, handleFocus, handleChange, placeholder, disabled,
}) => (
    <StyledInput disabled={ disabled } onChange={ handleChange } onFocus={ handleFocus } value={ value } placeholder={ placeholder } />
);

Input.propTypes = {
    /**
     * Current value of Input field
     */
    value: PropTypes.any,
    /**
     * Function triggered when Input is Focused, ie. user clicked into Input
     */
    handleFocus: PropTypes.func,
    /**
     * Function to handle value change on user input
     */
    handleChange: PropTypes.func,
    /**
     * String placeholder value, for when Input is empty
     */
    placeholder: PropTypes.string,
    /**
     * Whether select field is disabled or not
     */
    disabled: PropTypes.bool,
};

export { StyledInput };
export default Input;
