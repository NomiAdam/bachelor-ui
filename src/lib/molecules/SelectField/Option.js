import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { lightTheme } from '../../constants/theme';

const StyledOption = styled.div`
    width: 100%;
    display: inline-block;
    border: none;
	padding: 10px 20px;
	text-align: center;
	background-color: ${ ({ selected }) => (selected ? lightTheme.lightGrey : 'initial') };
    &:hover {
        background-color: ${ lightTheme.lightGrey };
        cursor: pointer;
    }
`;

const Option = ({
    value, onClick, label, selected,
}) => (
    <StyledOption onMouseDown={ onClick(value) } selected={ selected }>
        {label}
    </StyledOption>
);

Option.propTypes = {
    /**
     * String value of Option
     */
    value: PropTypes.any,
    /**
     * onMouseDown event handler, (value) => (e) =>
     */
    onClick: PropTypes.func.isRequired,
    /**
     * String representation of displayed value
     */
    label: PropTypes.string,
    /**
     * Boolean representation of currently selected item
     */
    selected: PropTypes.bool,
};

export { StyledOption };
export default Option;
