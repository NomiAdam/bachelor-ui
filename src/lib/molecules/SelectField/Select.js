import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import {
    map, length, equals, o,
} from 'ramda';
import Option from './Option';
import { darkTheme, lightTheme } from '../../constants/theme';
import DEFAULT_THEME from '../../utils/resolveTheme';

const StyledSelect = styled.div`
    ${ DEFAULT_THEME }
    overflow: hidden;
    display: ${ props => (props.visible ? 'flex' : 'none') }
    top: -50px;
    position: absolute;
    z-index: 999;
    width: ${ ({ width }) => width };
    min-height: 0px;
    max-height: 500px;
    justify-content: center;
    align-items: center;
    flex-direction: column;
    ${ ({ isEmpty }) => (isEmpty ? 'border: none;' : '') }
    font-size: 16px;
  	box-shadow: 5px 5px 15px ${ darkTheme.darkGrey };
`;

const doesItHaveOptions = o(equals(0), length);
const Select = ({
    visible, options, onClick, width, selected, backgroundColor,
}) => {
    const equalsSelected = equals(selected);
    const mapSuggestion = ({ value, label }) => (
        <Option value={ value } label={ label } onClick={ onClick } key={ value } selected={ equalsSelected(value) } />
    );
    const mappedOptions = map(mapSuggestion);
    return (
        <StyledSelect
            backgroundColor={ backgroundColor }
            visible={ visible }
            isEmpty={ doesItHaveOptions(options) }
            width={ width }
        >
            {visible && mappedOptions(options)}
        </StyledSelect>
    );
};

Select.propTypes = {
    /**
     * Whether select div is visible or not, is visible when Input is focused
     */
    visible: PropTypes.bool,
    /**
     * Array of possible values [ { value: 'Value', label: 'Label' }, ... ]
     */
    options: PropTypes.array,
    /**
     * Option onMouseDown handler
     */
    onClick: PropTypes.func,
    /**
     * String representation of select width
     */
    width: PropTypes.string,
    /**
     * String definition of currently selected item
     */
    selected: PropTypes.string,
    /**
   * Optional String color definition
   */
    backgroundColor: PropTypes.string,
};

Select.defaultProps = {
    options: [],
    backgroundColor: lightTheme.white,
};

export { StyledSelect };
export default Select;
