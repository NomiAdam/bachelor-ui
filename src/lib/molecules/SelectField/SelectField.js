/* eslint-disable default-case */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import {
    path,
    o,
    prop,
    head,
    propOr,
    propEq,
    when,
    findIndex,
    dec,
    add,
    nth,
} from 'ramda';
import { isNilOrEmptyString, isNotNil } from 'ramda-extension';
import { FaCaretDown } from 'react-icons/lib/fa';
import Input from './Input';
import Select from './Select';

const getInputValue = path(['target', 'value']);

const getFirstOptionsValue = o(propOr('', 'value'), head);
const getCurrentIndex = value => findIndex(propEq('value', value));
const getNextValue = when(isNotNil, prop('value'));

const StyledSelectField = styled.div`
  display: inline-block;
  width: ${ ({ width }) => width };
  position: relative;
  z-index: 9;
`;

const StyledCaretDown = styled(FaCaretDown)`
    position: absolute;
    top: 25%;
    right: 0;
    margin: 0 5px;
    font-size: 1.2rem;
    z-index: -999;
`;

class SelectField extends Component {
    state = {
        isFocused: false,
        selected: '',
    };

    componentDidMount() {
        const { value, options } = this.props;
        if (isNilOrEmptyString(value)) {
            this.setState({ selected: getFirstOptionsValue(options) });
        } else {
            this.setState({ selected: value });
        }
    }

    handleKeyPress = (e) => {
        const { selected } = this.state;
        const { options } = this.props;
        const index = getCurrentIndex(selected)(options);
        switch (e.key) {
        case 'ArrowUp': {
            const getValueOfNext = nth(dec(index, 1), options);
            this.setState({ selected: getNextValue(getValueOfNext) || selected });
        }
            break;
        case 'ArrowDown': {
            const getValueOfNext = nth(add(index, 1), options);
            this.setState({ selected: getNextValue(getValueOfNext) || selected });
        }
            break;
        case 'Enter':
            this.fireChange(selected);
            break;
        }
    };

    handleFocus = () => {
        this.setState({ isFocused: true });
    };

    handleBlur = () => {
        this.setState({ isFocused: false });
    };

    handleChange = (e) => {
        this.fireChange(getInputValue(e));
    };

    handleClick = value => () => {
        this.fireChange(value);
    };

    fireChange = (value) => {
        const { handleChange, disabled } = this.props;
        if (!disabled) {
            handleChange(value);
        }
    };

    render() {
        const {
            value, placeholder, options, width, disabled,
        } = this.props;
        const { isFocused, selected } = this.state;
        return (
            <StyledSelectField onBlur={ this.handleBlur } width={ width } onKeyDown={ this.handleKeyPress }>
                <Input
                    value={ value }
                    handleFocus={ this.handleFocus }
                    handleChange={ this.handleChange }
                    placeholder={ placeholder }
                    disabled={ disabled }
                />
                <Select
                    width={ width }
                    visible={ !disabled && isFocused }
                    options={ options }
                    selected={ selected }
                    onClick={ this.handleClick }
                />
                <StyledCaretDown />
            </StyledSelectField>
        );
    }
}

SelectField.propTypes = {
    /**
     * Array of possible values [ { value: 'Value', label: 'Label' }, ... ]
     */
    options: PropTypes.array,
    /**
     * Current value of Input component
     */
    value: PropTypes.any,
    /**
     * Function to handle Change on Input or Select
     */
    handleChange: PropTypes.func,
    /**
     * String definition of placeholder
     */
    placeholder: PropTypes.string,
    /**
     * String representation of preferred width of element
     */
    width: PropTypes.string,
    /**
     * Whether selectField is disabled or not
     */
    disabled: PropTypes.bool,
};

SelectField.defaultProps = {
    value: '',
    options: [],
    placeholder: '',
    width: '100%',
};

export { StyledSelectField };
export default SelectField;
