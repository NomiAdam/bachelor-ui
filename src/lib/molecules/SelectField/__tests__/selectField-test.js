/* eslint-disable no-undef */
import React from 'react';
import SelectField, {StyledSelectField} from '../SelectField';
import {shallow} from 'enzyme';

const mock = jest.fn();
const mockSuggestions = [{value: 'Value', label: 'Label'}];

describe('SelectField', () => {
    beforeEach(() => {
        jest.resetAllMocks();
    });
    it('should be defined', () => {
        expect(SelectField).toBeDefined();
    });
    it('should match the snapshot', () => {
        expect(shallow(<SelectField/>)).toMatchSnapshot();
    });
    it('should match the snapshot with options', () => {
        expect(shallow(<SelectField options={mockSuggestions}/>)).toMatchSnapshot();
    });
    it('should be of type StyledAutocomplete', () => {
        expect(shallow(<SelectField/>).type()).toBe(StyledSelectField);
    });
    it('should have default state', () => {
        expect(shallow(<SelectField/>).state()).toEqual(
            {
                isFocused: false,
                selected: '',
            }
        );
    });
    it('should not fire action on render', () => {
        shallow(<StyledSelectField handleChange={mock}/>);
        expect(mock).not.toHaveBeenCalled();
    });
});
