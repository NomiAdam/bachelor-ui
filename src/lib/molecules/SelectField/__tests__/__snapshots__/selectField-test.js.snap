// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectField should match the snapshot 1`] = `
<styled.div
  onBlur={[Function]}
  onKeyDown={[Function]}
  width="100%"
>
  <Input
    handleChange={[Function]}
    handleFocus={[Function]}
    placeholder=""
    value=""
  />
  <Select
    backgroundColor="#ffffff"
    onClick={[Function]}
    options={Array []}
    selected=""
    visible={false}
    width="100%"
  />
  <Styled(FaCaretDown) />
</styled.div>
`;

exports[`SelectField should match the snapshot with options 1`] = `
<styled.div
  onBlur={[Function]}
  onKeyDown={[Function]}
  width="100%"
>
  <Input
    handleChange={[Function]}
    handleFocus={[Function]}
    placeholder=""
    value=""
  />
  <Select
    backgroundColor="#ffffff"
    onClick={[Function]}
    options={
      Array [
        Object {
          "label": "Label",
          "value": "Value",
        },
      ]
    }
    selected="Value"
    visible={false}
    width="100%"
  />
  <Styled(FaCaretDown) />
</styled.div>
`;
